### PR TITLE
cuda: Load cuda shared libraries from linked/rpath/LD_LIBRARY_PATH

### PR DESCRIPTION
**Pull request by anustuvicl**
This PR adds the feature that the CUDA component loads the cuda toolkit version that is used by the application.

### CODE REVIEW

File: src/components/cuda/cupti_common.c [hunk 1]
```diff
@@ -356,8 +370,27 @@ void cuptic_disabled_reason_get(const char **pmsg)
...
+
 int cuptic_init(void)
 {
+    dl_iterate_phdr(callback_libcudart, NULL);
+    if (linked_cudart_path != NULL) {
+        LOGDBG("Linked cudart root: %s\n", linked_cudart_path);
+    }
+    else {
+        LOGDBG("Target application not linked with cuda runtime libraries.\n");
+    }
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:393]: I would wrap this in a function. E.g., `get_user_cudart_path`

File: src/components/cuda/tests/Makefile [hunk 2]
```diff
@@ -22,6 +22,8 @@ endif
 
 PAPI_FLAG = -DPAPI    # Comment this line for tests to run without PAPI profiling
 NVCFLAGS += -g -ccbin='$(CC)' $(PAPI_FLAG)
+# Uncomment this line to fix CUDA path in executable, comment to rely on LD_LIBRARY_PATH
+# NVCFLAGS += -Xlinker "-rpath,$(PAPI_CUDA_ROOT)/lib64"
```
Comment (gcongiu) [inline src/components/cuda/tests/Makefile:26]: What is this?

File: src/components/cuda/cupti_common.c [hunk 3]
```diff
@@ -356,8 +370,27 @@ void cuptic_disabled_reason_get(const char **pmsg)
     *pmsg = cuptic_disabled_reason_g;
 }
 
+static int callback_libcudart(struct dl_phdr_info *info, __attribute__((unused)) size_t size, __attribute__((unused)) void *data)
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:373]: `dl_iterate_phdr_cb` seems a better name.

File: src/components/cuda/cupti_common.c [hunk 4]
```diff
@@ -248,6 +258,10 @@ static int util_load_cuda_sym(void)
     papi_errno = load_cuda_sym();
     papi_errno += load_cudart_sym();
     papi_errno += load_cupti_common_sym();
+    if (linked_cudart_path) {
+        papi_free((void*) linked_cudart_path);
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:262]: Do you need to cast to void *?

File: src/components/cuda/cupti_common.c [hunk 5]
```diff
@@ -248,6 +258,10 @@ static int util_load_cuda_sym(void)
     papi_errno = load_cuda_sym();
     papi_errno += load_cudart_sym();
     papi_errno += load_cupti_common_sym();
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:260]: Unrelated to this PR. Please never do something like this. Always handle one function at a time.

**Review by gcongiu — CHANGES_REQUESTED** [review]

File: src/components/cuda/tests/Makefile (hunk 2 above)
Comment (anustuvicl) [reply src/components/cuda/tests/Makefile:26]: I made this a comment to have a discussion regarding whether we want this or not.

File: src/components/cuda/cupti_common.c (hunk 4 above)
Comment (anustuvicl) [reply src/components/cuda/cupti_common.c:276]: `linked_cudart_path` is a `const char *`. So freeing it raises a warning `ignores const qualifier`. Casting it to `void *` resolves the warning.

Comment (anustuvicl) [reply src/components/cuda/cupti_common.c:276]: This was also a bug. It shouldn't be freed in `util_load_cuda_sym`. Now it is freed in `cupti_shutdown`

File: src/components/cuda/cupti_common.c [hunk 6]
```diff
@@ -356,8 +380,31 @@ void cuptic_disabled_reason_get(const char **pmsg)
...
+static int get_user_cudart_path(void)
+{
+    dl_iterate_phdr(dl_iterate_phdr_cb, NULL);
+    if (linked_cudart_path != NULL) {
+        LOGDBG("Linked cudart root: %s\n", linked_cudart_path);
+    }
+    else {
+        LOGDBG("Target application not linked with cuda runtime libraries.\n");
+    }
+}
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:403]: This function should also return an integer. Return PAPI_OK

File: src/components/cuda/cupti_common.c [hunk 7]
```diff
@@ -356,8 +380,31 @@ void cuptic_disabled_reason_get(const char **pmsg)
...
+static int dl_iterate_phdr_cb(struct dl_phdr_info *info, __attribute__((unused)) size_t size, __attribute__((unused)) void *data)
+{
+    const char *library_name = "libcudart.so";
+    const char *library_path = info->dlpi_name;
+
+    if (library_path != NULL && strstr(library_path, library_name) != NULL) {
+        linked_cudart_path = strdup(dirname(dirname((char *) library_path)));
+    }
+
+    return 0;
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:392]: return PAPI_OK

File: src/components/cuda/tests/Makefile (hunk 2 above)
Comment (gcongiu) [reply src/components/cuda/tests/Makefile:26]: please remove. if you want to discuss we can do internally.

**Review by gcongiu — CHANGES_REQUESTED** [review]
Also rebase onto master. After you address the comments we can merge this.

File: src/components/cuda/tests/Makefile (hunk 2 above)
Comment (anustuvicl) [reply src/components/cuda/tests/Makefile]: Removed

**Comment by anustuvicl** [comment]
All requested changes applied and rebased.

File: src/components/cuda/cupti_common.c [hunk 8]
```diff
@@ -356,9 +375,39 @@ void cuptic_disabled_reason_get(const char **pmsg)
...
+
+static int get_user_cudart_path(void)
+{
+    dl_iterate_phdr(dl_iterate_phdr_cb, NULL);
+    if (linked_cudart_path != NULL) {
+        return PAPI_OK;
+    }
+    else {
+        return PAPI_EMISC;
+    }
```
Comment (gcongiu) [inline src/components/cuda/cupti_common.c:398]: This is a bit awkward as it is written. Can you change it to:
```
if (NULL == linked_cudart_path) {
    return PAPI_EMISC;
}
return PAPI_OK;
```

Comment (anustuvicl) [reply src/components/cuda/cupti_common.c]: Done

**Review by gcongiu — APPROVED** [review]
lets wait for the tests to finish running and then merge

**Pull request merged**